pupil_src/player/vis_circle.py

```diff
@@ -42,9 +42,22 @@ def update(self,frame,events):
         else:
             thickness = self.thickness
 
-        pts = [denormalize(pt['norm_pos'],frame.img.shape[:-1][::-1],flip_y=True) for pt in events.get('gaze_positions',[])]
-        for pt in pts:
-            transparent_circle(frame.img, pt, radius=self.radius, color=(self.b, self.g, self.r, self.a), thickness=thickness)
+        pts = [(
+            denormalize(pt['norm_pos'],frame.img.shape[:-1][::-1],flip_y=True),
+            pt['timestamp'])
+            for pt in events.get('gaze_positions',[])
+        ]
```

Comment:
Yes. Fixations are grouped by frame. The initial implementation would check if there is a fixation for the current frame and if yes render all pupil positions red. This led to the problem where saccades would be rendered red as well just because an fixation ended at the beginning of the frame.

Therefore it is needed to filter the pupil positions that are part of an fixations from those that are not — per frame.